storage/innobase/trx/trx0rseg.cc

```diff
@@ -60,29 +87,28 @@ trx_rseg_update_wsrep_checkpoint(
 	long long xid_seqno = wsrep_xid_seqno(xid);
 	const byte* xid_uuid = wsrep_xid_uuid(xid);
 
-	if (!memcmp(xid_uuid, wsrep_uuid, sizeof wsrep_uuid)) {
+	if (xid_seqno != -1
+	    && !memcmp(xid_uuid, wsrep_uuid, sizeof wsrep_uuid)) {
 		ut_ad(xid_seqno > wsrep_seqno);
 	} else {
 		memcpy(wsrep_uuid, xid_uuid, sizeof wsrep_uuid);
 	}
 	wsrep_seqno = xid_seqno;
 #endif /* UNIV_DEBUG */
+	trx_rseg_write_wsrep_checkpoint(rseg_header, xid, mtr);
+}
 
-	mlog_write_ulint(TRX_RSEG_WSREP_XID_FORMAT + rseg_header,
-			 uint32_t(xid->formatID),
-			 MLOG_4BYTES, mtr);
-
-	mlog_write_ulint(TRX_RSEG_WSREP_XID_GTRID_LEN + rseg_header,
-			 uint32_t(xid->gtrid_length),
-			 MLOG_4BYTES, mtr);
-
-	mlog_write_ulint(TRX_RSEG_WSREP_XID_BQUAL_LEN + rseg_header,
-			 uint32_t(xid->bqual_length),
-			 MLOG_4BYTES, mtr);
-
-	mlog_write_string(TRX_RSEG_WSREP_XID_DATA + rseg_header,
-			  reinterpret_cast<const byte*>(xid->data),
-			  XIDDATASIZE, mtr);
+/** Clear the WSREP XID information from rollback segment header.
+@param[in,out]	rseg_header	Rollback segment header
+@param[in,out]	mtr 		mini-transaction */
+static void
+trx_rseg_clear_wsrep_checkpoint(
+	trx_rsegf_t*	rseg_header,
+	mtr_t*		mtr)
+{
+	XID null_xid;
+	memset(&null_xid, 0, sizeof null_xid);
+	trx_rseg_write_wsrep_checkpoint(rseg_header, &null_xid, mtr);
```

Comment:
GCC 8 will emit a warning for `memset()` on a data object whose type defines member functions.
Do we really have to allocate such a constant from the stack and zero-fill it each time? Did you try the following:
```c++
	static const XID null_xid;
 	trx_rseg_write_wsrep_checkpoint(rseg_header, &null_xid, mtr);
```
If invoking this in the same mini-transaction in `trx_rseg_update_wsrep_checkpoint()` (like 10.4 does it) is generating too much redo log due to the lack of a `MLOG_MEMSET` record, then maybe we should write this zero-initialization as a special case, like this:
```c++
 	mlog_write_ulint(TRX_RSEG_WSREP_XID_FORMAT + rseg_header,
			 0, MLOG_4BYTES, mtr);
```
This should be sufficient to invalidate the WSREP XID, right? Or, perhaps it would suffice to merely write 0 to `TRX_SYS_WSREP_XID_MAGIC_N_FLD`?